### ISSUE TITLE
Specialized associated type not recognized as such

### ISSUE THREAD

**Issue by gkoz** [rust-lang/rust#22697]
In `fn fudge` rustc doesn't seem to recognize that the associated type `Wrapper::Inner` is specified for `Envelope<Foo>` to be `Foo` and makes me specify
```rust
fn fudge<T: Copy>(inner: <Envelope<Foo> as Wrapper>::Inner) -> T
```
instead of
```rust
fn fudge<T: Copy>(inner: Foo) -> T
```
```rust
#[derive(Copy)]
struct Foo;
#[derive(Copy)]
struct Bar;

trait Wrapper {
    type Inner;
    fn inner(&self) -> Self::Inner;
}

struct Envelope<S> {
    inner: S,
}

impl <S: Copy> Wrapper for Envelope<S> {
    type Inner = S;
    fn inner(&self) -> S {
        self.inner
    }
}

trait Fudge<T: Wrapper>: Wrapper {
    fn fudge(inner: <Self as Wrapper>::Inner) -> <T as Wrapper>::Inner;
}

impl Fudge<Envelope<Bar>> for Envelope<Foo> {
    fn fudge(inner: Foo) -> Bar {
        unsafe {
            ::std::mem::transmute(inner)
        }
    }
}

// error: mismatched types:
//  expected `<E<Foo> as Wrapper>::Inner`,
//     found `Foo`
// (expected associated type,
//     found struct `Foo`) [E0308]
fn fudge<T: Copy>(inner: Foo) -> T
//fn fudge<T: Copy>(inner: <Envelope<Foo> as Wrapper>::Inner) -> T
where Envelope<Foo>: Fudge<Envelope<T>> {
    <Envelope<_> as Fudge<Envelope<T>>>::fudge(inner)
}

fn main() {
    let f = Foo;
    let b: Bar = fudge(f);
}
```

**Comment by arielb1** [comment]
Fixed between 1.1-1.2.

**Comment by gkoz** [comment]
Thanks @arielb1 this compiles fine in the playpen on beta and nightly channels indeed.
